### PR TITLE
Feature/fix cn

### PR DESCRIPTION
**Pull request by Mandur**
# PR for issue #837

## What is being addressed

Fix some things in the China frequencies after the manual tests. 
Start removing the Dynamic region assignation (TryResolveRegion) and refactor tests.

### CODE REVIEW

File: Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino [hunk 1]
```diff
@@ -0,0 +1,113 @@
+
+#include <LoRaWan.h>
+// THOSE ARDUINO SAMPLE REQUIRES CHANGES IN THE ARDUINO CODE TO RUN DESCRIBED IN THE MAIN REPO
+// THE CONCENTRATOR IS EXPECTED TO HAVE THE FOLLOWING FREQUENCIES ACTIVATED
+// 498.3, 498.7, 498.9, 499.1, 499.3, 499.5, 499.7, 499.9
+// please refer to the repo documentation for further information
+
+//set to true to send confirmed data up messages
+bool confirmed = true;
+//application information, should be similar to what was provisiionned in the device twins
```
Comment (p-schuler) [inline Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino:10]:
```suggestion
//application information, should be similar to what was provisioned in the device twins
```

File: Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+char data[10];
+char buffer[256];
+int i = 0;
+int lastCall = 0;
+
+
+void setup(void)
+{
+  SerialUSB.begin(115200);
+  while (!SerialUSB);
```
Comment (p-schuler) [inline Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino:42]: That's a very tight loop, I assume we do this everywhere? Any way we could sleep for a bit before checking again?

File: Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    else
+      result = lora.transferPacket(data, 10);
+    i++;
+
+    if (result)
+    {
+      short length;
+      short rssi;
+
+      memset(buffer, 0, 256);
```
Comment (p-schuler) [inline Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino:94]: should we make the 256 a variable we can use? It's re-used in a few places. Same for the length of the data array?

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownstreamSender.cs [hunk 4]
```diff
@@ -85,8 +85,8 @@ private string Message(DownlinkPktFwdMessage message, Region region)
                 writer.WriteNumber("RxDelay", message.LnsRxDelay);
                 writer.WriteNumber("RX1DR", dataRate);
                 writer.WriteNumber("RX1Freq", (ulong)(message.Txpk.Freq * 1e6));
-                writer.WriteNumber("RX2DR", region.GetDefaultRX2ReceiveWindow().DataRate);
-                writer.WriteNumber("RX2Freq", (ulong)region.GetDefaultRX2ReceiveWindow().Frequency);
+                writer.WriteNumber("RX2DR", region.GetDefaultRX2ReceiveWindow(message.DeviceJoinInfo).DataRate);
```
Comment (p-schuler) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownstreamSender.cs:88]: We are getting the default RX2 receive window twice. I see that some of the regions have quite some logic to get to the window. Would it make sense to cache that information and only get it once?

File: Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino (hunk 2 above)
Comment (Mandur) [reply Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino:42]: What is your worry there? this code would trigger onlywhen we actively open the serial

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/MessageDispatcher.cs [hunk 5]
```diff
@@ -81,22 +80,14 @@ public void DispatchRequest(LoRaRequest request)
...
-                {
-                    // log is generated in Region factory
-                    // move here once V2 goes GA
-                    request.NotifyFailed(LoRaDeviceRequestFailedReason.InvalidRegion);
-                    return;
-                }
-
-                this.loraRegion = currentRegion;
+                // log is generated in Region factory
+                // move here once V2 goes GA
```
Comment (p-schuler) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/MessageDispatcher.cs:86]: how do we track this?

File: Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino (hunk 3 above)
Comment (atifaziz) [reply Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino:94]: Why not just use `sizeof(buffer)` here?

File: Tests/Common/MessageProcessorTestBase.cs [hunk 6]
```diff
@@ -71,10 +74,12 @@ public MessageProcessorTestBase()
             LoRaDeviceClient = new Mock<ILoRaDeviceClient>();
             this.cache = new MemoryCache(new MemoryCacheOptions() { ExpirationScanFrequency = TimeSpan.FromSeconds(5) });
             ConnectionManager = new LoRaDeviceClientConnectionManager(this.cache, NullLogger<LoRaDeviceClientConnectionManager>.Instance);
-
             var requestHandler = new DefaultLoRaDataRequestHandler(ServerConfiguration, FrameCounterUpdateStrategyProvider, new LoRaPayloadDecoder(NullLogger<LoRaPayloadDecoder>.Instance), deduplicationFactory, adrStrategyProvider, adrManagerFactory, functionBundlerProvider, NullLogger<DefaultLoRaDataRequestHandler>.Instance, meter: null);
-
             LoRaDeviceFactory = new TestLoRaDeviceFactory(ServerConfiguration, LoRaDeviceClient.Object, ConnectionManager, DeviceCache, requestHandler);
+
+            // By default we pick EU868 region.
+            Region = Enum.TryParse<LoRaRegionType>(Environment.GetEnvironmentVariable("REGION"), out var loraRegionType) ?
```
Comment (atifaziz) [inline Tests/Common/MessageProcessorTestBase.cs:81]: Consider making the parsing case insensitive:

```suggestion
            Region = Enum.TryParse<LoRaRegionType>(Environment.GetEnvironmentVariable("REGION"), ignoreCase: true, out var loraRegionType) ?
```

File: Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino (hunk 3 above)
Comment (p-schuler) [reply Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino:94]: much better! didn't know what was available in the arduino env.

File: Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino (hunk 2 above)
Comment (p-schuler) [reply Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino:42]: "we actively open the serial"? we do this here right? We wait for it to become available, no? If it is not, this is a very tight loop. I think we have this everywhere and it seems to work, so I assume it's ok?

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/MessageDispatcher.cs (hunk 5 above)
Comment (Mandur) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/MessageDispatcher.cs:86]: No idea, I did not write this, being honest I don't even know what this log mean. Do you?

File: Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino (hunk 2 above)
Comment (Mandur) [reply Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino:42]: Yeah, this is a copy from their sample, to be honest I don't know how much is the backoff/retry on this.

Comment (Mandur) [reply Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino:42]: If you see a problem with it I can make some research but this thing has always worked so far and it's in the sample repo

File: Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino (hunk 3 above)
Comment (Mandur) [reply Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino:94]: Changed in every arduino sample. Resolve if that is ok for you

File: Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino [hunk 7]
```diff
@@ -0,0 +1,113 @@
+
+#include <LoRaWan.h>
+// THOSE ARDUINO SAMPLE REQUIRES CHANGES IN THE ARDUINO CODE TO RUN DESCRIBED IN THE MAIN REPO
+// THE CONCENTRATOR IS EXPECTED TO HAVE THE FOLLOWING FREQUENCIES ACTIVATED
+// 498.3, 498.7, 498.9, 499.1, 499.3, 499.5, 499.7, 499.9
```
Comment (Mandur) [inline Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino:5]: This is RP1

Comment (Mandur) [reply Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino:5]: wrong

File: Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino (hunk 3 above)
Comment (p-schuler) [reply Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino]: I'm still seeing this:

memset(buffer, 0, 256);

did you push?

Comment (p-schuler) [reply Arduino/CN470RP2/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino]: looks like the browser was on a stale commit. Looks good.

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/MessageDispatcher.cs (hunk 5 above)
Comment (p-schuler) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/MessageDispatcher.cs:86]: This is a very old comment. Let's remove it I don't think it's accurate anymore.

**Review by p-schuler — APPROVED** [review]
LGTM

**Pull request merged**